cow_instrument/testing/IndexTranslatorTest.cpp

```diff
@@ -0,0 +1,229 @@
+  EXPECT_EQ(globalIndices.size(), 3);
+  EXPECT_EQ(globalIndices[0], std::vector<GlobalSpectrumIndex>({0}));
+  EXPECT_EQ(globalIndices[1], std::vector<GlobalSpectrumIndex>({4}));
+  EXPECT_EQ(globalIndices[2], std::vector<GlobalSpectrumIndex>({2}));
+
+  auto spectrumNumbers = t.spectrumNumbers(p3);
+  EXPECT_EQ(spectrumNumbers.size(), 3);
+  EXPECT_EQ(spectrumNumbers[0], std::vector<SpectrumNumber>({1}));
+  EXPECT_EQ(spectrumNumbers[1], std::vector<SpectrumNumber>({6}));
+  EXPECT_EQ(spectrumNumbers[2], std::vector<SpectrumNumber>({4}));
+}
+
+TEST(IndexTranslatorTest, test_split_2_to_3_partition_1) {
+  Partitioning p2(2);
+  IndexTranslator t(p2, 1, {1, 2, 4, 5, 6});
+  Partitioning p3(3);
+
+  auto indices = t.spectrumIndices(p3);
+  EXPECT_EQ(indices.size(), 3);
+  EXPECT_EQ(indices[0], std::vector<size_t>({1}));
+  EXPECT_EQ(indices[1], std::vector<size_t>({0}));
+  EXPECT_EQ(indices[2], std::vector<size_t>({}));
+
+  auto globalIndices = t.globalSpectrumIndices(p3);
+  EXPECT_EQ(globalIndices.size(), 3);
+  EXPECT_EQ(globalIndices[0], std::vector<GlobalSpectrumIndex>({3}));
+  EXPECT_EQ(globalIndices[1], std::vector<GlobalSpectrumIndex>({1}));
+  EXPECT_EQ(globalIndices[2], std::vector<GlobalSpectrumIndex>({}));
+
+  auto spectrumNumbers = t.spectrumNumbers(p3);
+  EXPECT_EQ(spectrumNumbers.size(), 3);
+  EXPECT_EQ(spectrumNumbers[0], std::vector<SpectrumNumber>({5}));
+  EXPECT_EQ(spectrumNumbers[1], std::vector<SpectrumNumber>({2}));
+  EXPECT_EQ(spectrumNumbers[2], std::vector<SpectrumNumber>({}));
+}
+
+TEST(IndexTranslatorTest, test_split_2_to_3_combine_results) {
+  // This tests mimics what would happen in an MPI redistribution
+  Partitioning p2(2);
+  std::vector<IndexTranslator> t;
+  t.emplace_back(
+      p2, 0, std::vector<SpectrumNumber>{1, 2, 4, 5, 6, 8, 9}); // partition 0
+  t.emplace_back(
+      p2, 1, std::vector<SpectrumNumber>{1, 2, 4, 5, 6, 8, 9}); // partition 1
+  Partitioning p3(3);
+
+  // Fake data object, could be anything.
+  using Data = size_t;
+
+  // Outermost vector is for target partitioning
+  // Middle vector is for input partitioning
+  // Innermost vector is spectrum index of input
+  std::vector<std::vector<
+      std::vector<std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data>>>>
+      receiveBuffer(p3.size());
+
+  // This for-loop mimics MPI ranks *before* redistribution
+  for (const auto partition : {0, 1}) {
+    // 1. Get local spectrum indices (used to extract data)
+    auto indices = t[partition].spectrumIndices(p3);
+
+    // 2. Use indices to get data from object that should be redistributed. Here
+    // we simply keep indices as fake data.
+    const auto &data = indices;
+
+    // 3. Tag data with GlobalSpectrumIndex and SpectrumNumber
+    auto globalIndices = t[partition].globalSpectrumIndices(p3);
+    auto spectrumNumbers = t[partition].spectrumNumbers(p3);
+    std::vector<
+        std::vector<std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data>>>
```

Comment:
Including spectrum numbers might be unnecessary. If we store the complete list of spectrum numbers in the translator having the global spectrum index should be enough.